0.3.4 breaks NAPI-RS

In https://github.com/napi-rs/napi-rs/blob/main/crates/napi/src/bindgen_runtime/mod.rs#L5, I re-export the `ctor::ctor` and uses it in the generted code (in user's project): https://github.com/napi-rs/napi-rs/blob/main/crates/backend/src/codegen/fn.rs#L705 to avoid requiring users to install `ctor`.

But in 0.3.4, the compiler complains:

```
error[E0433]: failed to resolve: could not find `ctor` in the list of imported crates
 --> src/lib.rs:3:1
  |
3 | #[napi]
  | ^^^^^^^ could not find `ctor` in the list of imported crates
```

--- a/ctor-proc-macro/Cargo.toml
+++ b/ctor-proc-macro/Cargo.toml
@@ -1,6 +1,6 @@
 [package]
 name = "ctor-proc-macro"
-version = "0.0.2"
+version = "0.0.4"
 authors = ["Matt Mastracci <matthew@mastracci.com>"]
 edition = "2018"
 description = "proc-macro support for the ctor crate"

--- a/ctor-proc-macro/src/lib.rs
+++ b/ctor-proc-macro/src/lib.rs
@@ -28,6 +28,36 @@ fn generate(
 ) -> TokenStream {
     let mut inner = TokenStream::new();
 
+    // Search for crate_path in attributes
+    let mut crate_path = None;
+    let mut tokens = attribute.clone().into_iter().peekable();
+    while let Some(token) = tokens.next() {
+        if let TokenTree::Ident(ident) = &token {
+            if ident.to_string() == "crate_path" {
+                // Look for =
+                if let Some(TokenTree::Punct(punct)) = tokens.next() {
+                    if punct.as_char() == '=' {
+                        // Collect tokens until comma or end
+                        let mut path = TokenStream::new();
+                        while let Some(token) = tokens.peek() {
+                            match token {
+                                TokenTree::Punct(p) if p.as_char() == ',' => {
+                                    tokens.next();
+                                    break;
+                                }
+                                _ => {
+                                    path.extend(std::iter::once(tokens.next().unwrap()));
+                                }
+                            }
+                        }
+                        crate_path = Some(path);
+                        break;
+                    }
+                }
+            }
+        }
+    }
+
     #[cfg(feature = "used_linker")]
     inner.extend([
         TokenTree::Punct(Punct::new('#', Spacing::Alone)),
@@ -91,21 +121,28 @@ fn generate(
 
     inner.extend(item);
 
-    TokenStream::from_iter([
-        TokenTree::Punct(Punct::new(':', Spacing::Joint)),
-        TokenTree::Punct(Punct::new(':', Spacing::Alone)),
-        TokenTree::Ident(Ident::new(macro_crate, Span::call_site())),
+    let mut invoke = crate_path.unwrap_or_else(|| {
+        TokenStream::from_iter([
+            TokenTree::Punct(Punct::new(':', Spacing::Joint)),
+            TokenTree::Punct(Punct::new(':', Spacing::Alone)),
+            TokenTree::Ident(Ident::new(macro_crate, Span::call_site())),
+        ])
+    });
+
+    invoke.extend([
         TokenTree::Punct(Punct::new(':', Spacing::Joint)),
         TokenTree::Punct(Punct::new(':', Spacing::Alone)),
         TokenTree::Ident(Ident::new("__support", Span::call_site())),
         TokenTree::Punct(Punct::new(':', Spacing::Joint)),
         TokenTree::Punct(Punct::new(':', Spacing::Alone)),
         TokenTree::Ident(Ident::new(
-            &format!("{}_parse", macro_crate),
+            &format!("{}_parse", macro_type),
             Span::call_site(),
         )),
         TokenTree::Punct(Punct::new('!', Spacing::Alone)),
         TokenTree::Group(Group::new(Delimiter::Parenthesis, inner)),
         TokenTree::Punct(Punct::new(';', Spacing::Alone)),
-    ])
+    ]);
+
+    invoke
 }

--- a/ctor/Cargo.toml
+++ b/ctor/Cargo.toml
@@ -20,7 +20,7 @@ used_linker = ["ctor-proc-macro/used_linker"]
 __warn_on_missing_unsafe = ["ctor-proc-macro/__warn_on_missing_unsafe"]
 
 [dependencies]
-ctor-proc-macro = "=0.0.2"
+ctor-proc-macro = "=0.0.4"
 
 [dev-dependencies]
 libc-print = "0.1.20"

--- a/ctor/src/lib.rs
+++ b/ctor/src/lib.rs
@@ -20,9 +20,9 @@ pub mod __support {
     pub use crate::__ctor_entry as ctor_entry;
     pub use crate::__ctor_link_section as ctor_link_section;
     pub use crate::__ctor_link_section_attr as ctor_link_section_attr;
-    #[doc(hidden)]
     pub use crate::__ctor_parse as ctor_parse;
     pub use crate::__dtor_entry as dtor_entry;
+    pub use crate::__dtor_parse as dtor_parse;
     pub use crate::__if_has_feature as if_has_feature;
     pub use crate::__if_unsafe as if_unsafe;
     pub use crate::__unify_features as unify_features;
@@ -30,25 +30,66 @@ pub mod __support {
 
 mod macros;
 
-// Code note:
-
-// You might wonder why we don't use `__attribute__((destructor))`/etc for
-// dtor. Unfortunately mingw doesn't appear to properly support section-based
-// hooks for shutdown, ie:
-
-// https://github.com/Alexpux/mingw-w64/blob/d0d7f784833bbb0b2d279310ddc6afb52fe47a46/mingw-w64-crt/crt/crtdll.c
-
-// In addition, OSX has removed support for section-based shutdown hooks after
-// warning about it for a number of years:
-
-// https://reviews.llvm.org/D45578
+/// Declarative forms of the `#[ctor]` and `#[dtor]` macros.
+///
+/// The declarative forms wrap and parse a proc_macro-like syntax like so, and
+/// are identical in expansion to the undecorated procedural macros. The
+/// declarative forms support the same attribute parameters as the procedural
+/// macros.
+///
+/// ```rust
+/// # mod test { use ctor::*; use libc_print::*;
+/// ctor::declarative::ctor! {
+///   #[ctor]
+///   fn foo() {
+///     libc_println!("Hello, world!");
+///   }
+/// }
+/// # }
+///
+/// // ... the above is identical to:
+///
+/// # mod test_2 { use ctor::*; use libc_print::*;
+/// #[ctor]
+/// fn foo() {
+///   libc_println!("Hello, world!");
+/// }
+/// # }
+/// ```
+pub mod declarative {
+    #[doc(inline)]
+    pub use crate::__support::ctor_parse as ctor;
+    #[doc(inline)]
+    pub use crate::__support::dtor_parse as dtor;
+}
 
 /// Marks a function or static variable as a library/executable constructor.
-/// This uses OS-specific linker sections to call a specific function at
-/// load time.
+/// This uses OS-specific linker sections to call a specific function at load
+/// time.
 ///
-/// Multiple startup functions/statics are supported, but the invocation order is not
-/// guaranteed.
+/// # Important notes
+///
+/// Rust does not make any guarantees about stdlib support for life-before or
+/// life-after main. This means that the `ctor` crate may not work as expected
+/// in some cases, such as when used in an `async` runtime or making use of
+/// stdlib services.
+///
+/// Multiple startup functions/statics are supported, but the invocation order
+/// is not guaranteed.
+///
+/// The `ctor` crate assumes it is available as a direct dependency, with
+/// `extern crate ctor`. If you re-export `ctor` items as part of your crate,
+/// you can use the `crate_path` parameter to redirect the macro's output to the
+/// correct crate.
+///
+/// # Attribute parameters
+///
+///  - `crate_path = ::path::to::ctor::crate`: The path to the `ctor` crate
+///    containing the support macros. If you re-export `ctor` items as part of
+///    your crate, you can use this to redirect the macro's output to the
+///    correct crate.
+///  - `used(linker)`: Use the linker to load the constructor.
+///  - `link_section = "section"`: The section to place the constructor in.
 ///
 /// # Examples
 ///
@@ -61,7 +102,8 @@ mod macros;
 /// use libc_print::std_name::println;
 ///
 /// #[ctor]
-/// fn foo() {
+/// unsafe fn foo() {
+///   // Using libc_print which is safe in `#[ctor]`
 ///   println!("Hello, world!");
 /// }
 ///
@@ -81,7 +123,7 @@ mod macros;
 /// static INITED: AtomicBool = AtomicBool::new(false);
 ///
 /// #[ctor]
-/// fn foo() {
+/// unsafe fn set_inited() {
 ///   INITED.store(true, Ordering::SeqCst);
 /// }
 /// # }
@@ -112,25 +154,87 @@ mod macros;
 ///
 /// # Details
 ///
-/// The `#[ctor]` macro makes use of linker sections to ensure that a
-/// function is run at startup time.
+/// The `#[ctor]` macro makes use of linker sections to ensure that a function
+/// is run at startup time.
+///
+/// ```rust
+/// # #![cfg_attr(feature="used_linker", feature(used_with_arg))]
+/// # extern crate ctor;
+/// # mod test {
+/// # use ctor::*;
+/// #[ctor]
+///
+/// unsafe fn my_init_fn() {
+///   /* ... */
+/// }
+/// # }
+/// ```
 ///
 /// The above example translates into the following Rust code (approximately):
 ///
-///```rust
+/// ```rust
+/// # fn my_init_fn() {}
 /// #[used]
-/// #[cfg_attr(any(target_os = "linux", target_os = "android"), link_section = ".init_array")]
-/// #[cfg_attr(target_os = "freebsd", link_section = ".init_array")]
-/// #[cfg_attr(target_os = "netbsd", link_section = ".init_array")]
-/// #[cfg_attr(target_os = "openbsd", link_section = ".init_array")]
-/// #[cfg_attr(target_os = "illumos", link_section = ".init_array")]
+/// #[cfg_attr(target_os = "linux", link_section = ".init_array")]
 /// #[cfg_attr(target_vendor = "apple", link_section = "__DATA,__mod_init_func")]
 /// #[cfg_attr(target_os = "windows", link_section = ".CRT$XCU")]
-/// static FOO: extern fn() = {
-///   #[cfg_attr(any(target_os = "linux", target_os = "android"), link_section = ".text.startup")]
-///   extern fn foo() { /* ... */ };
-///   foo
+/// /* ... other platforms elided ... */
+/// static INIT_FN: extern fn() = {
+///     extern fn init_fn() { my_init_fn(); };
+///     init_fn
 /// };
+/// ```
+///
+/// For `static` items, the macro generates a `std::sync::OnceLock` that is
+/// initialized at startup time.
+///
+/// ```rust
+/// # extern crate ctor;
+/// # mod test {
+/// # use ctor::*;
+/// # use std::collections::HashMap;
+/// #[ctor]
+/// static FOO: HashMap<u32, String> = unsafe {
+///   let mut m = HashMap::new();
+///   for i in 0..100 {
+///     m.insert(i, format!("x*100={}", i*100));
+///   }
+///   m
+/// };
+/// # }
+/// ```
+///
+/// The above example translates into the following Rust code (approximately),
+/// which eagerly initializes the `HashMap` inside a `OnceLock` at startup time:
+///
+/// ```rust
+/// # extern crate ctor;
+/// # mod test {
+/// # use ctor::ctor;
+/// # use std::collections::HashMap;
+/// static FOO: FooStatic = FooStatic { value: ::std::sync::OnceLock::new() };
+/// struct FooStatic {
+///   value: ::std::sync::OnceLock<HashMap<u32, String>>,
+/// }
+///
+/// impl ::std::ops::Deref for FooStatic {
+///   type Target = HashMap<u32, String>;
+///   fn deref(&self) -> &Self::Target {
+///     self.value.get_or_init(|| unsafe {
+///       let mut m = HashMap::new();
+///       for i in 0..100 {
+///         m.insert(i, format!("x*100={}", i*100));
+///       }
+///       m
+///     })
+///   }
+/// }
+///
+/// #[ctor]
+/// unsafe fn init_foo_ctor() {
+///   _ = &*FOO;
+/// }
+/// # }
 /// ```
 pub use ctor_proc_macro::ctor;
 
@@ -139,9 +243,6 @@ pub use ctor_proc_macro::ctor;
 ///
 /// Multiple shutdown functions are supported, but the invocation order is not
 /// guaranteed.
-///
-/// `sys_common::at_exit` is usually a better solution for shutdown handling, as
-/// it allows you to use `stdout` in your handlers.
 ///
 /// ```rust
 /// # #![cfg_attr(feature="used_linker", feature(used_with_arg))]

--- a/dtor/src/lib.rs
+++ b/dtor/src/lib.rs
@@ -4,9 +4,8 @@ pub mod __support {
     pub use crate::__ctor_entry as ctor_entry;
     pub use crate::__ctor_link_section as ctor_link_section;
     pub use crate::__ctor_link_section_attr as ctor_link_section_attr;
-    #[doc(hidden)]
-    pub use crate::__ctor_parse as dtor_parse;
     pub use crate::__dtor_entry as dtor_entry;
+    pub use crate::__dtor_parse as dtor_parse;
     pub use crate::__if_has_feature as if_has_feature;
     pub use crate::__if_unsafe as if_unsafe;
 }
@@ -34,3 +33,34 @@ mod macros;
 /// }
 /// ```
 pub use dtor_proc_macro::dtor;
+
+/// Declarative forms of the `#[dtor]` macro.
+///
+/// The declarative forms wrap and parse a proc_macro-like syntax like so, and
+/// are identical in expansion to the undecorated procedural macros. The
+/// declarative forms support the same attribute parameters as the procedural
+/// macros.
+///
+/// ```rust
+/// # mod test { use dtor::*; use libc_print::*;
+/// dtor::declarative::dtor! {
+///   #[dtor]
+///   fn foo() {
+///     libc_println!("Goodbye, world!");
+///   }
+/// }
+/// # }
+///
+/// // ... the above is identical to:
+///
+/// # mod test_2 { use dtor::*; use libc_print::*;
+/// #[dtor]
+/// fn foo() {
+///   libc_println!("Goodbye, world!");
+/// }
+/// # }
+/// ```
+pub mod declarative {
+    #[doc(inline)]
+    pub use crate::__support::dtor_parse as dtor;
+}

--- a/shared/src/macros/mod.rs
+++ b/shared/src/macros/mod.rs
@@ -1,8 +1,25 @@
-/// Macro entry point. At this point it's either:
+/// Parse a `#[ctor]`-annotated item as if it were a proc-macro.
 ///
-/// `#[ctor] (pub) (unsafe) fn IDENT`
-/// `#[dtor] (pub) (unsafe) fn IDENT`
-/// `#[ctor] (pub) static IDENT`
+/// This macro supports both the `fn` and `static` forms of the `#[ctor]`
+/// attribute, including attribute parameters.
+///
+/// ```rust
+/// # #[cfg(any())] mod test {
+/// ctor! {
+///   #[ctor(link_section = "...")]
+///   unsafe fn foo() { /* ... */ }
+/// }
+///
+/// ctor! {
+///   #[ctor(link_section = "...")]
+///   static FOO: std::collections::HashMap<u32, String> = unsafe {
+///     let mut m = std::collections::HashMap::new();
+///     m.insert(1, "foo".to_string());
+///     m
+///   };
+/// }
+/// # }
+/// ```
 #[doc(hidden)]
 #[macro_export]
 macro_rules! __ctor_parse {
@@ -21,7 +38,20 @@ macro_rules! __ctor_parse {
     ($( #[feature($fname:ident)] )* #[ctor $(($($meta:tt)*))?] $(#[$imeta:meta])* static $($item:tt)*) => {
         $crate::__support::unify_features!(next=$crate::__support::ctor_entry, meta=[$($($meta)*)?], features=[$($fname,)*], imeta=$(#[$imeta])*, vis=[], item=static $($item)*);
     };
+}
 
+/// Parse a `#[dtor]`-annotated item as if it were a proc-macro.
+///
+/// ```rust
+/// # #[cfg(any())] mod test {
+/// dtor! {
+///   #[dtor]
+///   unsafe fn foo() { /* ... */ }
+/// }
+/// # }
+#[doc(hidden)]
+#[macro_export]
+macro_rules! __dtor_parse {
     ($( #[feature($fname:ident)] )* #[dtor $(($($meta:tt)*))?] $(#[$imeta:meta])* pub ( $($extra:tt)* ) $($item:tt)*) => {
         $crate::__support::dtor_entry!(meta=[$($($meta)*)?], features=[$($fname,)*], imeta=$(#[$imeta])*, vis=[pub($($extra)*)], item=$($item)*);
     };
@@ -39,6 +69,7 @@ macro_rules! __ctor_parse {
 /// Extract #[ctor/dtor] attribute parameters and turn them into features.
 ///
 /// Supported attributes:
+///
 /// - `used(linker)` -> feature: `used_linker`
 /// - `link_section = ...` -> feature: `(link_section = ...)`
 #[doc(hidden)]
@@ -49,6 +80,9 @@ macro_rules! __unify_features {
     };
     (next=$next_macro:path, meta=[link_section = $section:tt $(, $($meta:tt)* )?], features=[$($features:tt,)*], $($rest:tt)*) => {
         $crate::__support::unify_features!(next=$next_macro, meta=[$($($meta)*)?], features=[(link_section=$section),$($features,)*], $($rest)*);
+    };
+    (next=$next_macro:path, meta=[crate_path = $path:path $(, $($meta:tt)* )?], features=[$($features:tt,)*], $($rest:tt)*) => {
+        $crate::__support::unify_features!(next=$next_macro, meta=[$($($meta)*)?], features=[(crate_path=$path),$($features,)*], $($rest)*);
     };
     (next=$next_macro:path, meta=[$unknown_meta:meta $($meta:tt)*], features=[$($features:tt,)*], $($rest:tt)*) => {
         compile_error!(concat!("Unknown attribute parameter: ", stringify!($unknown_meta)));
@@ -224,6 +258,19 @@ macro_rules! __ctor_entry {
         }
     };
 }
+
+// Code note:
+
+// You might wonder why we don't use `__attribute__((destructor))`/etc for
+// dtor. Unfortunately mingw doesn't appear to properly support section-based
+// hooks for shutdown, ie:
+
+// https://github.com/Alexpux/mingw-w64/blob/d0d7f784833bbb0b2d279310ddc6afb52fe47a46/mingw-w64-crt/crt/crtdll.c
+
+// In addition, OSX has removed support for section-based shutdown hooks after
+// warning about it for a number of years:
+
+// https://reviews.llvm.org/D45578
 
 #[doc(hidden)]
 #[macro_export]

